dynamically add warning label to iop when needed

dterrahe opened the pull request:
This will allow dt_iop_set_module_trouble_message to be used in any iop, without explicitly needing to set up a widget for it in gui_init.

I haven't tested it with dt_iop_have_required_input_format because I don't know how to trigger a warning there.

The logic in _set_trouble_message was not changed, even though it seems that only the first error will get written to stderr. Which makes sense if we want to avoid repeating identical messages, but also means that if the message changes, it will get swallowed as well.

Also fixes a couple of semi-unrelated g_strdup leaks.

ralfbrown commented on src/develop/imageop.c:
```diff
@@ -1360,10 +1362,26 @@ void dt_iop_set_module_in_trouble(dt_iop_module_t *module, const gboolean state)
+
+  if(module && module->widget)
+  {
+    GtkWidget *iopw = gtk_widget_get_parent(module->widget);
```
You should not unconditionally create the widget -- set_module_trouble_message gets called with NULL message every time the module's process() function is invoked to clear any prior error condition (unless there is in fact [still] an error to report).  So on the vast majority of calls, your code will be creating and then immediately destroying the widget.

dterrahe replied:
Makes a lot of sense! I think this is now fixed.

dterrahe commented:
Added a red highlight in the warning icon. Removed the icon from the warning label widget, since that already uses a red background for highlighting. Also simplified \" -> ' in some existing markup messages elsewhere (sorry for the pollution of this PR).

I'm a little worried about ambiguities in the interfaces and interactions between dt_iop_set_module_in_trouble and dt_iop_set_module_trouble_message. They seem to mostly overlap but not completely. At the moment, setting a message will also set the trouble flag, but clearing the trouble flag does _not_ remove the message. Could this be simplified by _only_ allowing setting the message and flag at the same time, via dt_iop_set_module_trouble_message (and clearing by passing NULL)? Setting the flag without informing the user of the reason does not seem too useful? So I'm suggesting basically to integrate dt_iop_set_module_in_trouble  into dt_iop_set_module_trouble_message or otherwise to drop the public interface (and rename to _iop_set_module_in_trouble). There's also some duplication in checking if the module is enabled before setting trouble flag and then _again_ checking before showing the trouble icon in the module header. I'd say only doing the latter should be sufficient and might simplify what needs to happen when enabling the module (although that does seem to work correctly at the moment).

I also don't understand (and might not have supported correctly) the intent behind stderr printing. Currently there seems to be only a single trouble condition per module, but the situation where a message was already displayed (has_trouble == TRUE) but then a _different_ message is set would (intentionally?) not lead to the new message being printed.  That check
`if((!module || !module->has_trouble) && (stderr_message || !label_widget))`
needs to be adjusted to take into account that the label_widget would only be set up later, but I left it for now because I don't understand the intent.

dterrahe commented:
There's also some support for module == NULL (printing a "?" to stderr). Not sure if that is useful. If anything, I'd check and throw an error at the start of the function; it is specifically meant to show a warning in the context of a function. It should not be used for generic stderr printing.

ralfbrown commented on src/develop/imageop.c:
```diff
@@ -1363,56 +1363,57 @@ static void _set_trouble_message(dt_iop_module_t *const module, const char* cons
+    iopw = gtk_widget_get_parent(module->widget);
     GList *children = gtk_container_get_children(GTK_CONTAINER(iopw));
     label_widget = g_list_nth_data(children, 0);
     g_list_free(children);
```
Can you use module->warning_label to cache the widget pointer and avoid all of this work?  99.99+% of calls will have trouble_message==NULL and the module not currently in trouble.  And this function will be called multiple times per pixelpipe run.

dterrahe replied:
If the interface can be disambiguated so that `has_trouble == TRUE` is always in sync with `trouble_msg != NULL` then has_trouble can be used to determine if there is a widget that needs to be removed if there is no (new) trouble_msg. I'd rather clarify/simplify the interface than add cache fields to paper over the inefficiencies it creates.

ralfbrown replied:
That would work as well.  Let me see whether merging set_module_in_trouble into set_module_trouble_message causes any problems elsewhere.

dterrahe replied:
Hi @ralfbrown were you still looking at this or was this completely resolved in another PR? I lost track a little, sorry.

ralfbrown commented:
> I'm a little worried about ambiguities in the interfaces and interactions between dt_iop_set_module_in_trouble and dt_iop_set_module_trouble_message.

I'm fine with merging them as suggested.  I was trying not to change other stuff.

> I also don't understand (and might not have supported correctly) the intent behind stderr printing. Currently there seems to be only a single trouble condition per module, but the situation where a message was already displayed (has_trouble == TRUE) but then a _different_ message is set would (intentionally?) not lead to the new message being printed.

Probably could be improved by storing a copy of the previous message so we know when it changes.  But since we expect the trouble messages to be **extremely** rare (except for the white-balance ones from the White Balance and Color Calibration iops), it didn't seem like a high priority.  The rarity of trouble messages is why I've been commenting on not unconditionally doing Gtk stuff on every invocation -- we want to minimize overhead except when the warning gets turned on or off (and in those rare cases, an extra 100ms wouldn't matter).

TurboGit commented:
I've lost track of this work. Reading the comment again it seems like a general enhancement that could be useful, and instead of having ad-hoc support developed in different modules I think that it is a good idea to implement this.

Anyway, before reviewing/testing the code this needs a conflict resolution.

ralfbrown commented:
I haven't yet gotten around to looking into the merging of functions I mentioned above, but expect to be able to do so this weekend.

dterrahe commented:
Cleaned up (removed) the interface that allowed directly setting the "trouble" flag without setting or removing the message. Now, setting the flag requires providing the accompanying message and the ui code is simplified. @ralfbrown please let me know if you see any problems.

Since I touched some markups (to turn the warning triangle red) I also changed other markups to use ' rather than \"

This made me look again at the colorlabels toast code and realise that an int color code was converted into a string and then painstakingly converted back again. Much simpler now. Not really sure why we need both LABELS_ICONS LABELS_COLORICONS, since both have the markup and they look identical to me, but I've tried not to "break" it. I also changed thumbtable to use the CSS colors.

@TurboGit Unless @ralfbrown raises an issue, I think this is ready for testing.

ralfbrown commented:
I didn't see any obvious issues during a quick glance at the current revision.

TurboGit commented:
Thanks !

TurboGit commented:
Of course issues are manifesting only after merging something :) Well it is not always reproducible, but maybe you'll have more chance than me.

- select an image in modern workflow
- clear history
- open white balance and change setting
- the while balance & color calibration should have a warning

In some case I don't have the warning on color calibration. Only when I go into the module and open/close it the red warning is displayed. This seems to manifest more the first time an image is edited.

TurboGit commented:
I have a reproducer now:
- select an image and create a mismatch between color calibration and white balance by change while balance parameter
- go to lighttable
- return to darkrrom
- see that while balance warning is there
- see that color calibration warning is not visible
- open/close color calibration to restore the warning

Seems like a refresh issue.
While entering the darkroom you may see a flickering of the warning in the color calibration as if it was displayed and then hidden.

dterrahe commented:
Do you think that is a regression from this PR? I had the feeling there was some odd but not always reproducable behavior before. This patch should not really impact _when_ the warning is shown, just _how_. I don't really understand the interaction/locking between the modules involved and at what point the header should be refreshed. If @ralfbrown has a better grasp hopefully he can have a look this weekend. If not, let me know and I'll see if I can figure it out.

TurboGit commented:
> Do you think that is a regression from this PR?

I would think so as I have not seen this refresh issue before. But of course I may be mistaken or it could be due to something else committed recently.
